tests/test_h2_encoder.c

```diff
@@ -0,0 +1,360 @@
+    static int s_test_##NAME(struct aws_allocator *allocator, void *ctx)
+
+/* Run the given frame's encoder and check that it outputs the expected bytes */
+static int s_encode(
+    struct aws_allocator *allocator,
+    struct aws_h2_frame_header *frame_header,
+    const uint8_t *expected,
+    size_t expected_size) {
+
+    struct aws_h2_frame_encoder encoder;
+    ASSERT_SUCCESS(aws_h2_frame_encoder_init(&encoder, allocator));
+
+    struct aws_byte_buf buffer;
+    /* Allocate more room than necessary, easier to debug the full output than a failed aws_h2_encode() call */
+    ASSERT_SUCCESS(aws_byte_buf_init(&buffer, allocator, expected_size * 2));
+
+    ASSERT_SUCCESS(aws_h2_encode(&encoder, frame_header, &buffer));
+    ASSERT_BIN_ARRAYS_EQUALS(expected, expected_size, buffer.buffer, buffer.len);
+
+    aws_byte_buf_clean_up(&buffer);
+    aws_h2_frame_encoder_clean_up(&encoder);
+    return AWS_OP_SUCCESS;
+}
+
+TEST_CASE(h2_encoder_data) {
+    (void)ctx;
+
+    struct aws_h2_frame_data frame;
+    ASSERT_SUCCESS(aws_h2_frame_data_init(&frame, allocator));
+    frame.header.stream_id = 0x76543210;
+    frame.end_stream = true;
+    frame.pad_length = 2;
+    frame.data = aws_byte_cursor_from_c_str("hello");
+
+    /* clang-format off */
+    uint8_t expected[] = {
+        0x00, 0x00, 0x08,           /* Length (24) */
+        AWS_H2_FRAME_T_DATA,        /* Type (8) */
+        AWS_H2_FRAME_F_END_STREAM | AWS_H2_FRAME_F_PADDED, /* Flags (8) */
+        0x76, 0x54, 0x32, 0x10,     /* Reserved (1) | Stream Identifier (31) */
+        /* DATA */
+        0x02,                       /* Pad Length? (8) */
+        'h', 'e', 'l', 'l', 'o',    /* Data (*) */
+        0x00, 0x00                  /* Padding (*) */
+    };
+    /* clang-format on */
+
+    ASSERT_SUCCESS(s_encode(allocator, &frame.header, expected, sizeof(expected)));
+    aws_h2_frame_data_clean_up(&frame);
+    return AWS_OP_SUCCESS;
+}
+
+TEST_CASE(h2_encoder_headers) {
+    (void)ctx;
+
+    struct aws_h2_frame_headers frame;
+    ASSERT_SUCCESS(aws_h2_frame_headers_init(&frame, allocator));
+    frame.header.stream_id = 0x76543210;
+    frame.end_headers = true;
+    frame.end_stream = true;
+    frame.pad_length = 2;
+    frame.has_priority = true;
+    frame.priority.stream_dependency_exclusive = true;
+    frame.priority.stream_dependency = 0x01234567;
+    frame.priority.weight = 9;
+
+    /* Intentionally leaving header block fragment empty. Header block encoding is tested elsewhere */
+
+    /* clang-format off */
+    uint8_t expected[] = {
+        0x00, 0x00, 0x08,           /* Length (24) */
+        AWS_H2_FRAME_T_HEADERS,     /* Type (8) */
+        AWS_H2_FRAME_F_END_STREAM | AWS_H2_FRAME_F_END_HEADERS | AWS_H2_FRAME_F_PADDED | AWS_H2_FRAME_F_PRIORITY, /* Flags (8) */
+        0x76, 0x54, 0x32, 0x10,     /* Reserved (1) | Stream Identifier (31) */
+        /* HEADERS */
+        0x02,                       /* Pad Length? (8) */
```

Comment:
IS this a ?

Reply:
I copied this from the spec, but in the spec it's part of a big ascii-art picture. I see how it's confusing here by itself. It means that it's only present when a specific FLAG is used earlier. I'll find a better way to illustrate this

Reply:
```
0x02,                       /* Pad Length (8)                           - F_PADDED */
0x81, 0x23, 0x45, 0x67,     /* Exclusive (1) | Stream Dependency (31)   - F_PRIORITY*/
0x09,                       /* Weight (8)                               - F_PRIORITY */
                            /* Header Block Fragment (*) */
0x00, 0x00                  /* Padding (*)                              - F_PADDED */
```